Join paths correctly

kleinfreund opened the pull request:
This pull request fixes #231 by using `TemplatePath.join` in a couple of new places, creating more robust data and includes directory paths.

I splitted the pull request into two main commits with the following changes.

**Join paths together correctly with TemplatePath utility**:

- replaces instances of path concatenation with the use of the available `TemplatePath.normalize` method.
- fixes a few tests which assumed a leading `"./"` segment. That’s gone after normalization.

**Separate normalize and join methods**:

- There are now two methods, `TemplatePath.normalize` and `TemplatePath.join`. First of all, this mirrors what Node.js does. Secondly, it makes the code more clear. It’s odd to call a normalizing function with multiple parameters. Reading a call `TemplatePath.normalize('path', 'test')` doesn’t make it very clear that the paths will be joined.
- I also separated the test cases for these two functions.

Phew, that’s it. *sweatin’* :sweat_drops: 

---

Once this gets merged (well, if it gets merged), I’d like to discuss whether Eleventy should stop users from escaping Eleventy’s project directory (e.g. via double-dot segments in the includes directory). I opened #368 for this.

zachleat commented:
Whoa, awesome! Gimme a bit to review this but on first glance it looks amazing—thank you!

kleinfreund commented:
@zachleat It just occured to me that the name `TemplatePath` is not adequate for the two methods because its scope is too narrow.

What do you think about a `EleventyPath.join` and `EleventyPath.normalize`? Especially in the scope of #368, these methods might be applied throughout the project, not only when templating paths are concerned.

zachleat commented:
@kleinfreund yes, I like that idea. However I’d like to review this ASAP as I’m shooting to release 0.7.0 this week. We can do some renames later, although I’m not sure if the rename will be worth the effort?

kleinfreund commented:
@zachleat I will look into it later. I think it’s worth it, yes. It makes the code more readable without adding any cost.

kleinfreund commented on src/TemplateConfig.js:
```diff
@@ -67,8 +67,10 @@ class TemplateConfig {
-      TemplatePath.getWorkingDir() + "/" + localProjectConfigPath
+    let path = TemplatePath.join(
+      TemplatePath.getWorkingDir(),
+      "/",
```
Yes, indeed. I pushed a new commit addressing the two mentioned issues.

kleinfreund commented:
@zachleat the leading `./`s are stripped by normalize-path’s `normalize` method. That’s one of the differences to Node.js’ `path.normalize`. Regarding file-watching, maybe we should add them again after normalizing?

I added the slash changes and rebased against the master branch; hence, the commits were force-pushed again.

kleinfreund commented:
@zachleat Ah, thanks for the heads-up. I think I stumbled over it a couple of times without paying further attention. Cool. :)
